### PR TITLE
Reduced cohomology  groups + Eilenberg-Steenrod axioms for cohomology

### PR DESCRIPTION
**Pull request by aljungstrom**
This PR contains:

1. Reduced cohomology (group structure)
2. A def of the ES axioms for (ordinary) reduced cohomology with binary additivity.
This is based on Cavallo's MSc thesis and Buchholtz/Favonia (2018)
3. A proof that our definition of reduced cohomology satisfies the axioms
4. Some minor things in AbGroup

### CODE REVIEW

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    A : Type ℓ
+    B : Type ℓ'
+
+cofib : (f : A → B) → Type _
+cofib f = Pushout (λ _ → tt) f
+
+cfcod : (f : A → B) → B → cofib f
+cfcod f = inr
+
+suspFun : (f : A → B) → Susp A → Susp B
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:41]: This should go into the Susp files. You could also use it to make https://github.com/agda/cubical/blob/master/Cubical/HITs/Susp/Properties.agda#L48 nicer

File: Cubical/ZCohomology/GroupStructure.agda [hunk 2]
```diff
@@ -406,10 +544,45 @@ coHomGr n A = coHom n A , coHomGrnA
...
+0g (snd (coHomRedGrDir n A)) = 0ₕ∙ n
+GroupStr._+_ (snd (coHomRedGrDir n A)) = _+ₕ∙_ {n = n}
+- snd (coHomRedGrDir n A) = -ₕ∙_ {n = n}
+isGroup (snd (coHomRedGrDir n A)) = helper
+  where
+  abstract
+    helper : IsGroup (0ₕ∙ n) (_+ₕ∙_ {n = n}) (-ₕ∙_ {n = n})
+    helper = makeIsGroup § (assocₕ∙ n) (rUnitₕ∙ n) (lUnitₕ∙ n) (rCancelₕ∙ n) (lCancelₕ∙ n)
+
+coHomRedGroupDir : (n : ℕ) (A : Pointed ℓ) → AbGroup {ℓ}
```
Comment (mortberg) [inline Cubical/ZCohomology/GroupStructure.agda:568]: Shouldn't the name have "Ab" in it? 

Also, why do you have `coHomRedGrDir` at all when you can get it from this lemma by projection?

Comment (mortberg) [reply Cubical/ZCohomology/GroupStructure.agda:568]: Actually ignore my comment about "Ab", the name is fine

File: Cubical/ZCohomology/Properties.agda [hunk 3]
```diff
@@ -199,6 +200,47 @@ leftInv (Iso-coHom-coHomRed {A = A , a} n) =
...
+coHomGr≅coHomRedGr : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
+                  → GroupEquiv (coHomRedGrDir (suc n) A) (coHomGr (suc n) (typ A))
+GroupEquiv.eq (coHomGr≅coHomRedGr n A) = isoToEquiv (Iso-coHom-coHomRed n)
+GroupEquiv.isHom (coHomGr≅coHomRedGr n A) = +∙≡+ n
+
+coHomGr≡coHomRedGr : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
+                  → coHomRedGrDir (suc n) A ≡ coHomGr (suc n) (typ A)
+coHomGr≡coHomRedGr n A = uaGroup (coHomGr≅coHomRedGr n A)
+
+coHomGroup≡coHomRedGroup' : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
```
Comment (mortberg) [inline Cubical/ZCohomology/Properties.agda:227]: Shouldn't this be local to the lemma below?

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    Suspension : Σ[ F ∈ ((n : Int) {A : Pointed ℓ} → AbGroupEquiv (H (sucInt n) (Susp (typ A) , north)) (H n A)) ]
+                   ({A B : Pointed ℓ} (f : A →∙ B) (n : Int)
+               → fun (f* (sucInt n) (suspFun (fst f) , refl)) ∘ invEq (eq (F n {A = B}))
+                ≡ invEq (eq (F n {A = A})) ∘ fun (f* n f))
+    Exactness : {A B : Pointed ℓ}  (f : A →∙ B) → (n :  Int)
+              → ((x : _) → isInKer _ _ (f* n {A = A} {B = B} f) x
+                          → isInIm _ _ (f* n {A = B} {B = _ , inr (pt B)} (cfcod (fst f) , refl)) x)
+               × ((x : _) → isInIm _ _ (f* n {A = B} {B = _ , inr (pt B)} (cfcod (fst f) , refl)) x
+                          → isInKer _ _ (f* n {A = A} {B = B} f) x)
+    Dimension : (n : ℕ) → isContr (fst (H (pos (suc n)) (Lift (Bool) , lift true)))
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:62]: Why not just say `n != 0`? That should lead to fewer things to prove later?

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+record isCohomTheory {ℓ : Level} (H : (n : Int) → Pointed ℓ → AbGroup {ℓ'}) : Type (ℓ-suc (ℓ-max ℓ ℓ'))
+  where
+  field
+    f* : (n : Int) → contravar (H n)
+    Suspension : Σ[ F ∈ ((n : Int) {A : Pointed ℓ} → AbGroupEquiv (H (sucInt n) (Susp (typ A) , north)) (H n A)) ]
+                   ({A B : Pointed ℓ} (f : A →∙ B) (n : Int)
+               → fun (f* (sucInt n) (suspFun (fst f) , refl)) ∘ invEq (eq (F n {A = B}))
+                ≡ invEq (eq (F n {A = A})) ∘ fun (f* n f))
+    Exactness : {A B : Pointed ℓ}  (f : A →∙ B) → (n :  Int)
+              → ((x : _) → isInKer _ _ (f* n {A = A} {B = B} f) x
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:58]: Are these propositions? If so then it would be nice to express this using a path instead

File: Cubical/Homotopy/EilenbergSteenrod.agda (hunk 4 above)
Comment (mortberg) [reply Cubical/Homotopy/EilenbergSteenrod.agda:62]: Also `Lift (Bool) , lift true` should be defined outside

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+cfcod : (f : A → B) → B → cofib f
+cfcod f = inr
+
+suspFun : (f : A → B) → Susp A → Susp B
+suspFun f north = north
+suspFun f south = south
+suspFun f (merid a i) = merid (f a) i
+
+contravar :  (H : Pointed ℓ → AbGroup {ℓ'}) → Type _
+contravar {ℓ = ℓ} H = {A B : Pointed ℓ} (f : A →∙ B) → AbGroupHom (H B) (H A)
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:47]: Why not inline this in the type of `f^*`? It doesn't seem like a very useful definition on its own and it doesn't pop up anywhere else

File: Cubical/ZCohomology/Properties.agda (hunk 3 above)
Comment (aljungstrom) [reply Cubical/ZCohomology/Properties.agda:227]: I'm using it twice below, so that won't work really. If I understand what you're saying correctly, that is:-)

File: Cubical/Homotopy/EilenbergSteenrod.agda (hunk 5 above)
Comment (aljungstrom) [reply Cubical/Homotopy/EilenbergSteenrod.agda:58]: Ker and Im would be sets (as total spaces over isInker and isInIm), and I guess this is really what we want. But it's an easy fix anyway, so I'll fix this.

**Comment by aljungstrom** [comment]
Everything fixed apart from
https://github.com/agda/cubical/pull/481#discussion_r544416173

File: Cubical/Algebra/Group/Morphism.agda [hunk 7]
```diff
@@ -51,6 +51,12 @@ isInIm G H ϕ h = ∃[ g ∈ ⟨ G ⟩ ] ϕ .fun g ≡ h
 isInKer : (G : Group {ℓ}) (H : Group {ℓ'}) → GroupHom G H → ⟨ G ⟩ → Type ℓ'
 isInKer G H ϕ g = ϕ .fun g ≡ 0g (snd H)
 
+Ker : (G : Group {ℓ}) (H : Group {ℓ'}) → GroupHom G H → Type _
```
Comment (mortberg) [inline Cubical/Algebra/Group/Morphism.agda:54]: Do we need G and H to be explicit?

File: Cubical/Algebra/Group/Morphism.agda [hunk 8]
```diff
@@ -51,6 +51,12 @@ isInIm G H ϕ h = ∃[ g ∈ ⟨ G ⟩ ] ϕ .fun g ≡ h
...
+Ker G H ϕ = Σ[ x ∈ ⟨ G ⟩ ] isInKer G H ϕ x
+
+Im : (G : Group {ℓ}) (H : Group {ℓ'}) → GroupHom G H → Type _
```
Comment (mortberg) [inline Cubical/Algebra/Group/Morphism.agda:57]: Same here

File: Cubical/HITs/Pushout/Base.agda [hunk 9]
```diff
@@ -15,6 +15,13 @@ data Pushout {ℓ ℓ' ℓ''} {A : Type ℓ} {B : Type ℓ'} {C : Type ℓ''}
   inr : C → Pushout f g
   push : (a : A) → inl (f a) ≡ inr (g a)
 
+-- cofiber
```
Comment (mortberg) [inline Cubical/HITs/Pushout/Base.agda:18]: Add a comment that this is equivalent to the mapping cone HIT

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 10]
```diff
@@ -0,0 +1,46 @@
+{-# OPTIONS --cubical --no-import-sorts --safe #-}
+
+module Cubical.Homotopy.EilenbergSteenrod where
+
+{-
+This module contains the Eilenberg-Steenrod axioms for ordinary reduced cohomology theories with binary additivity.
+The axioms are based on the ones given in Cavallo's MSc thesis (https://www.cs.cmu.edu/~ecavallo/works/thesis15.pdf) and Buchholtz/Favonia (2018)
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:7]: Long line

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  Boolℓ = Lift Bool , lift true
+  field
+    f* : (n : Int) → {A B : Pointed ℓ} (f : A →∙ B) → AbGroupHom (H n B) (H n A)
+    Suspension : Σ[ F ∈ ((n : Int) {A : Pointed ℓ} → AbGroupEquiv (H (sucInt n) (Susp (typ A) , north)) (H n A)) ]
+                   ({A B : Pointed ℓ} (f : A →∙ B) (n : Int)
+               → fun (f* (sucInt n) (suspFun (fst f) , refl)) ∘ invEq (eq (F n {A = B}))
+                ≡ invEq (eq (F n {A = A})) ∘ fun (f* n f))
+    Exactness : {A B : Pointed ℓ}  (f : A →∙ B) (n :  Int)
+              → Ker _ _ (f* n {A = A} {B = B} f)
+               ≡ Im _ _ (f* n {A = B} {B = _ , inr (pt B)} (cfcod (fst f) , refl))
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:44]: Looks like the first two arguments of Ker and Im should indeed be implicit.

Aso, do you need to give A and B in f*? It makes things quite hard to parse...

File: Cubical/Homotopy/EilenbergSteenrod.agda [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+open import Cubical.Data.Bool
+open import Cubical.Data.Sigma
+open import Cubical.Data.Int
+
+open import Cubical.Algebra.Group
+open import Cubical.Algebra.AbGroup
+open GroupEquiv
+open GroupHom
+
+record isCohomTheory {ℓ ℓ' : Level} (H : (n : Int) → Pointed ℓ → AbGroup {ℓ'}) : Type (ℓ-suc (ℓ-max ℓ ℓ'))
```
Comment (mortberg) [inline Cubical/Homotopy/EilenbergSteenrod.agda:32]: I would drop the `is` as this is not a property but a structure

File: Cubical/ZCohomology/Properties.agda [hunk 13]
```diff
@@ -312,6 +351,42 @@ private
...
+    trElim (λ _ → isProp→isOfHLevelSuc (3 + n) (isPropIsOfHLevel (3 + n)))
+      (sphereToPropElim _
+        (λ _ → (isPropIsOfHLevel (3 + n))) (isOfHLevelTrunc (3 + n)))
+
+  Code-add' : {n : ℕ} (x : _) → Code n ∣ north ∣ → Code n ∣ x ∣ → Code n ∣ x ∣
+  Code-add' {n = n} north = _+ₖ_
+  Code-add' {n = n} south = _+ₖ_
+  Code-add' {n = n} (merid a i) = helper n a i
+    where
+    help : (n : ℕ) → (x y a : S₊ (suc n)) → transport (λ i → Code n ∣ north ∣ → Code n ∣ merid a i ∣ → Code n ∣ merid a i ∣) (_+ₖ_) ∣ x ∣ ∣ y ∣ ≡ ∣ x ∣ +ₖ ∣ y ∣
```
Comment (mortberg) [inline Cubical/ZCohomology/Properties.agda:365]: Long line

File: Cubical/ZCohomology/Properties.agda [hunk 14]
```diff
@@ -199,6 +200,44 @@ leftInv (Iso-coHom-coHomRed {A = A , a} n) =
...
+                   (coHomRed (suc n) A , _+ₕ∙_)
+                   (isoToEquiv (invIso (Iso-coHom-coHomRed n)))
+                   (homhelp n A))
+
+coHomRedGroup : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ) → AbGroup {ℓ}
+coHomRedGroup zero A = coHomRedGroupDir zero A
+coHomRedGroup (suc n) A = coHomGroup≡coHomRedGroup' n A i0
+
+abstract
+  coHomGroup≡coHomRedGroup : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
```
Comment (mortberg) [inline Cubical/ZCohomology/Properties.agda:237]: Maybe you should add a comment to the top of this file that the equality of reduced and unreduced cohomology can be found in it?

**Review by mortberg — CHANGES_REQUESTED** [review]
A few more comments :-)

File: Cubical/Homotopy/EilenbergSteenrod.agda (hunk 11 above)
Comment (aljungstrom) [reply Cubical/Homotopy/EilenbergSteenrod.agda:44]: Yeah, agda gets confused about the choice of base point in A and B otherwise:-/

**Comment by aljungstrom** [comment]
> A few more comments :-)

Fixed

File: Cubical/Homotopy/EilenbergSteenrod.agda (hunk 11 above)
Comment (mortberg) [reply Cubical/Homotopy/EilenbergSteenrod.agda]: Does it help if you change the definition of pointed functions to 
```
_→∙_ : ∀ {ℓ ℓ'} → (A : Pointed ℓ) (B : Pointed ℓ') → Type (ℓ-max ℓ ℓ')
(A , ptA) →∙ (B , ptB) = Σ[ f ∈ (A → B) ] f ptA ≡ ptB
```
in `Cubical.Foundations.Pointed.Base`?

**Comment by mortberg** [comment]
Great! Merging as soon as Travis is done (as usual please remind me if I forget :-) )

File: Cubical/ZCohomology/Properties.agda [hunk 15]
```diff
@@ -199,6 +213,44 @@ leftInv (Iso-coHom-coHomRed {A = A , a} n) =
...
+coHomGr≅coHomRedGr : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
+                  → GroupEquiv (coHomRedGrDir (suc n) A) (coHomGr (suc n) (typ A))
+GroupEquiv.eq (coHomGr≅coHomRedGr n A) = isoToEquiv (Iso-coHom-coHomRed n)
+GroupEquiv.isHom (coHomGr≅coHomRedGr n A) = +∙≡+ n
+
+private
+  coHomGroup≡coHomRedGroup' : ∀ {ℓ} (n : ℕ) (A : Pointed ℓ)
+                          → _ ≡ coHomGroup (suc n) (typ A)
+  coHomGroup≡coHomRedGroup' n A =
+    sym (InducedAbGroupPath (coHomGroup (suc n) (typ A))
```
Comment (mortberg) [inline Cubical/ZCohomology/Properties.agda:240]: @cangiuli Here's the nice use of the SIP for AbGroup that I mentioned the other day. @aljungstrom had to do some hacks to make Agda accept the definition quickly, but apart from this it's a nice application. I should try to figure out why it's so slow to use without hacks

Comment (mortberg) [reply Cubical/ZCohomology/Properties.agda:240]: @mzeuner and @ecavallo might also enjoy this

Comment (mortberg) [reply Cubical/ZCohomology/Properties.agda:240]: Ignoring the hacks we're showing that the reduced and unreduced Z-cohomology groups are equal as abelian groups. To make our lives easier we provide a direct definition of addition on reduced cohomology and it's then only the axioms that get transported over from the unreduced groups

**Review by mortberg — APPROVED** [review]

File: Cubical/ZCohomology/Properties.agda (hunk 15 above)
Comment (aljungstrom) [reply Cubical/ZCohomology/Properties.agda:240]: Oh wups, I just realised I don't need the hack anymore. Defining AbGroup->Group by copatterns was enough for Agda to be happy. I'll make a new wee PR

Comment (mortberg) [reply Cubical/ZCohomology/Properties.agda:240]: Wow, awesome!

Comment (aljungstrom) [reply Cubical/ZCohomology/Properties.agda:240]: Ah sorry, I think I was to trigger-happy. Agda is doing something very odd right now... I'll open a draft PR and show you
